File: cpp/047.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        permute(nums, 0, res);
        return res;
    }

    void permute(vector<int> nums, int start, vector<vector<int>>& res) {
        if (start == nums.size()) {
            res.push_back(nums);
            return;
        }
        int size = nums.size();
        for (int i = start; i < size; ++i) {
            if (i != start && nums[i] == nums[start]) 
                continue;
            swap(nums[i], nums[start]);
            permute(nums, start + 1, res);
        }
    }
};

File: go/047.go
func permuteUnique(nums []int) [][]int {
    sort.Ints(nums)
    var res [][]int
    help(nums, 0, &res)
    return res
}

func help(nums []int, start int, res *[][]int) {
    if start == len(nums) {
        temp := make([]int, len(nums))
        copy(temp, nums)
        *res = append(*res, temp)
        return
    }
    size := len(nums)
    for i := start; i < size; i++ {
        if i != start && nums[i] == nums[start] {
            continue
        }
        nums[i], nums[start] = nums[start], nums[i]
        tmp := make([]int, len(nums))
        copy(tmp, nums)
        help(tmp, start+1, res)
    }
}
